projects/CenterNet2/centernet2/custom_gaussian_focal_loss.py

```diff
@@ -0,0 +1,114 @@
+import torch.nn as nn
+import torch
+
+from mmdet.registry import MODELS
+from mmdet.models.losses.utils import weighted_loss
+
+
+@weighted_loss
+def custom_gaussian_focal_loss(
+    pred, 
+    gaussian_target,
+    pos_inds = None,
+    alpha: float = -1,
+    beta: float = 4,
+    gamma: float = 2,
+    sigmoid_clamp: float = 1e-4,
+    ignore_high_fp: float = -1.,
+    ):
+    """`Focal Loss <https://arxiv.org/abs/1708.02002>`_ for targets in gaussian
+    distribution.
+    Args:
+        pred (torch.Tensor): The prediction.
+        gaussian_target (torch.Tensor): The learning target of the prediction
+            in gaussian distribution.
+        alpha (float, optional): A balanced form for Focal Loss.
+            Defaults to 2.0.
+        gamma (float, optional): The gamma for calculating the modulating
+            factor. Defaults to 4.0.
+    """
+    pred = torch.clamp(pred.sigmoid_(), min=sigmoid_clamp, max=1-sigmoid_clamp)
+    neg_weights = torch.pow(1 - gaussian_target, beta)
+    pos_pred = pred[pos_inds] # N
+    pos_loss = torch.log(pos_pred) * torch.pow(1 - pos_pred, gamma)
+    neg_loss = torch.log(1 - pred) * torch.pow(pred, gamma) * neg_weights
+    if ignore_high_fp > 0:
+        not_high_fp = (pred < ignore_high_fp).float()
+        neg_loss = not_high_fp * neg_loss
+
+    pos_loss = - pos_loss.sum()
+    neg_loss = - neg_loss.sum()
+
+    if alpha >= 0:
+        pos_loss = alpha * pos_loss
+        neg_loss = (1 - alpha) * neg_loss
+
+    return pos_loss + neg_loss
+
+@MODELS.register_module()
+class CustomGaussianFocalLoss(nn.Module):
```

Comment:
Can we directly use the GaussianFocalLoss implemented in MMDet for CornerNet?

Reply:
CN2's GFL should support `ignore_high_fp`
https://github.com/xingyizhou/CenterNet2/blob/8745e012e4dbdf560ac2f27e0b771d4907ad4aaf/centernet/modeling/layers/heatmap_focal_loss.py#L34